File: clojure/rna-transcription/src/rna_transcription.clj
(ns rna-transcription
  (:require [clojure.string :as str]))

(defn transcribe-nucleotide
  "Given a nucleotide return its complement."
  [n]
  (case n
    "G" "C"
    "C" "G"
    "T" "A"
    "A" "U"
    (throw (AssertionError. "Provided invalid DNA strand."))))

(defn to-rna
  "Given a DNA strand, return its RNA complement."
  [dna]
  (apply str (map transcribe-nucleotide (str/split dna #""))))

File: zig/rna-transcription/rna_transcription.zig
const mem = @import("std").mem;

pub const RnaError = error{Invalid};

fn transcribeNucleotide(nucleotide: u8) RnaError!u8 {
    return switch (nucleotide) {
        'G' => 'C',
        'C' => 'G',
        'T' => 'A',
        'A' => 'U',
        else => return RnaError.Invalid,
    };
}

pub fn toRna(allocator: mem.Allocator, dna: []const u8) (RnaError || mem.Allocator.Error)![]const u8 {
    var rna = try allocator.alloc(u8, dna.len);
    for (dna) |n, i| rna[i] = try transcribeNucleotide(n);
    return rna;
}
